ktoolbox/common.py

```diff
@@ -557,30 +566,365 @@ def structparse_check_strdict(arg: Any, yamlpath: str) -> dict[str, Any]:
 def structparse_check_empty_dict(vdict: dict[str, Any], yamlpath: str) -> None:
     length = len(vdict)
     if length == 1:
-        raise ValueError(f'"{yamlpath}": unknown key "{list(vdict)[0]}"')
+        raise ValueError(f'"{yamlpath}": unknown key {repr(list(vdict)[0])}')
     if length > 1:
         raise ValueError(f'"{yamlpath}": unknown keys {list(vdict)}')
 
 
-def structparse_check_and_pop_name(
-    vdict: dict[str, Any], yamlpath: str, *, required: bool = False
-) -> Optional[str]:
-    name = vdict.pop("name", None)
-    if name is None:
-        if required:
-            raise ValueError(f'"{yamlpath}.name": mandatory key missing')
-        return None
-    if not isinstance(name, str):
-        raise ValueError(f'"{yamlpath}.name": expects a string but got {name}')
-    return name
+@dataclass(frozen=True)
+class StructParseVarg:
+    vdict: dict[str, Any]
+    yamlpath: str
+    check_empty: bool = dataclasses.field(default=True, init=False)
 
+    def for_key(self, key: str) -> tuple[dict[str, Any], str, str]:
+        return self.vdict, self.yamlpath, key
 
-def structparse_check_and_pop_name_required(
-    vdict: dict[str, Any], yamlpath: str
-) -> str:
-    return typing.cast(
-        str, structparse_check_and_pop_name(vdict, yamlpath, required=True)
+    def for_name(self, key: str = "name") -> tuple[dict[str, Any], str, str]:
+        return self.vdict, self.yamlpath, key
+
+    def skip_check_empty(self) -> None:
+        object.__setattr__(self, "check_empty", False)
+
+
+@contextlib.contextmanager
+def structparse_with_strdict(
```

Comment:
Can we have some function headers that tell the user how to use this function as docs? This is especially helpful if we want ktoolbox to be a commonly used across projects. People need some kind of docs for these functions and their intended use.